libraries/GCS_MAVLink/GCS_Rally.cpp

```diff
@@ -101,3 +102,73 @@ void GCS_MAVLINK::handle_common_rally_message(mavlink_message_t *msg)
     }
 }
 
+MAV_MISSION_RESULT MissionItemProtocol_Rally::get_item(const GCS_MAVLINK &_link,
+                                                       const mavlink_message_t &msg,
+                                                       const mavlink_mission_request_int_t &packet,
+                                                       mavlink_mission_item_int_t &ret_packet)
+{
+    RallyLocation rallypoint;
+
+    if (!rally.get_rally_point_with_index(packet.seq, rallypoint)) {
+        return MAV_MISSION_INVALID_SEQUENCE;
+    }
+
+    ret_packet.command = MAV_CMD_NAV_RALLY_POINT;
+    ret_packet.x = rallypoint.lat;
+    ret_packet.y = rallypoint.lng;
+    ret_packet.z = rallypoint.alt;
+
+    return MAV_MISSION_ACCEPTED;
+}
+
+void MissionItemProtocol_Rally::set_count(const mavlink_mission_count_t &packet)
+{
+    expected_item_count = packet.count;
+}
+void MissionItemProtocol_Rally::timeout()
+{
+    link->send_text(MAV_SEVERITY_WARNING, "Rally upload timeout");
+    expected_item_count = -1;
+}
+
+void MissionItemProtocol_Rally::convert_MISSION_ITEM_INT_to_RallyLocation(const mavlink_mission_item_int_t &cmd, RallyLocation &ret) const
+{
+    ret.lat = cmd.x;
+    ret.lng = cmd.y;
+    ret.alt = constrain_int32(cmd.z, INT16_MIN, INT16_MAX);
+}
+
+MAV_MISSION_RESULT MissionItemProtocol_Rally::replace_item(const mavlink_mission_item_int_t &cmd)
+{
+    RallyLocation rallyloc;
+    convert_MISSION_ITEM_INT_to_RallyLocation(cmd, rallyloc);
+    if (!rally.set_rally_point_with_index(cmd.seq, rallyloc)) {
+        return MAV_MISSION_ERROR;
+    }
+    return MAV_MISSION_ACCEPTED;
+}
+
+MAV_MISSION_RESULT MissionItemProtocol_Rally::append_item(const mavlink_mission_item_int_t &cmd)
+{
+    RallyLocation rallyloc;
+    convert_MISSION_ITEM_INT_to_RallyLocation(cmd, rallyloc);
+    const uint8_t current_total = rally.get_rally_total();
+    if (!rally.set_rally_point_with_index(current_total, rallyloc)) {
+        return MAV_MISSION_ERROR;
+    }
+    rally.set_rally_total(current_total+1);
+    return MAV_MISSION_ACCEPTED;
+}
+
+void MissionItemProtocol_Rally::complete(const GCS_MAVLINK &_link)
+{
+    _link.send_text(MAV_SEVERITY_INFO, "Rally points received");
+    if (expected_item_count >= 0) {
+        // note that expected_item_count will only be >0 if we started
+        // the transfer with a mission-item-count, not a
+        // write-partial-list.
+        rally.set_rally_total(expected_item_count);
```

Comment:
Why do we need to change it here? Shouldn't the adding items have fixed the total already?

Reply:
Not for Rally.  Rally never appends, only replaces, due to the expected-item-count.

This means not having to change the Rally interface to do `append(...)`.

Into the future we will be able to create a bitmask of items yet to be uploaded.  We will need something like expected-item-count on the waypoints at that time.  Re-requesting waypoints will be done by looking at he first set (or unset) but in the bitmask.